File: app/src/main/java/com/github/overpass/gather/auth/register/RegistrationFragment.java
package com.github.overpass.gather.auth.register;

import android.content.Context;

import com.github.overpass.gather.base.BaseFragment;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;

public abstract class RegistrationFragment<VM extends ViewModel> extends BaseFragment<VM> {

    protected RegistrationController registrationController;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof RegistrationController) {
            registrationController = (RegistrationController) context;
        } else {
            throw new RuntimeException(context + " must implement "
                    + RegistrationController.class.getSimpleName());
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        registrationController = null;
    }
}

File: app/src/main/java/by/overpass/gather/ui/auth/register/RegistrationFragment.kt
package by.overpass.gather.ui.auth.register

import android.content.Context
import android.util.Log
import androidx.lifecycle.ViewModel
import by.overpass.gather.ui.base.BaseFragmentKt

abstract class RegistrationFragment<VM : ViewModel> : BaseFragmentKt<VM>() {

    protected var registrationController: RegistrationController? = null

    protected fun getInitialStep(): Int {
        return registrationController?.getInitialStep() ?: 0
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is RegistrationController) {
            registrationController = context
        } else {
            Log.d(TAG, context.toString() + " must implement "
                    + RegistrationController::class.java.simpleName)
        }
    }

    override fun onDetach() {
        super.onDetach()
        registrationController = null
    }

    companion object {

        private const val TAG = "RegistrationFragment"
    }
}
